firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/UpdateApkClientTest.java

```diff
@@ -0,0 +1,169 @@
+    FirebaseAppDistributionException e =
+        (FirebaseAppDistributionException) updateTask.getException();
+    assertEquals(Constants.ErrorMessages.NETWORK_ERROR, e.getMessage());
+    assertEquals(FirebaseAppDistributionException.Status.DOWNLOAD_FAILURE, e.getErrorCode());
+  }
+
+  @Test
+  public void updateApk_whenInstallSuccessful_setsResult() throws Exception {
+    List<UpdateProgress> progressEvents = new ArrayList<>();
+    UpdateTaskImpl updateTask = new UpdateTaskImpl();
+    doReturn(Tasks.forResult(mockFile)).when(updateApkClient).downloadApk(TEST_URL, updateTask);
+    updateTask.addOnProgressListener(progressEvents::add);
+
+    updateApkClient.updateApk(updateTask, TEST_URL, activity);
+    latch.await(1000, TimeUnit.MILLISECONDS);
+    updateApkClient.setInstallationResult(RESULT_OK);
+    assertEquals(1, progressEvents.size());
+    assertEquals(
+        UpdateProgress.builder()
+            .setApkBytesDownloaded(TEST_FILE_LENGTH)
+            .setApkFileTotalBytes(TEST_FILE_LENGTH)
+            .setUpdateStatus(UpdateStatus.DOWNLOADED)
+            .build(),
+        progressEvents.get(0));
+    assertTrue(updateTask.isSuccessful());
+  }
+
+  @Test
+  public void updateApk_whenInstallCancelled_setsError() throws Exception {
+    UpdateTaskImpl updateTask = new UpdateTaskImpl();
+    doReturn(Tasks.forResult(mockFile)).when(updateApkClient).downloadApk(TEST_URL, updateTask);
+
+    updateApkClient.updateApk(updateTask, TEST_URL, activity);
+    latch.await(1000, TimeUnit.MILLISECONDS);
+    updateApkClient.setInstallationResult(RESULT_CANCELED);
+
+    assertFalse(updateTask.isSuccessful());
+    assertTrue(updateTask.getException() instanceof FirebaseAppDistributionException);
+    FirebaseAppDistributionException e =
+        (FirebaseAppDistributionException) updateTask.getException();
+    assertEquals(Constants.ErrorMessages.UPDATE_CANCELED, e.getMessage());
+    assertEquals(FirebaseAppDistributionException.Status.INSTALLATION_CANCELED, e.getErrorCode());
+  }
+
+  @Test
+  public void updateApk_whenInstallFailed_setsError() throws Exception {
+    UpdateTaskImpl updateTask = new UpdateTaskImpl();
+    doReturn(Tasks.forResult(mockFile)).when(updateApkClient).downloadApk(TEST_URL, updateTask);
+
+    updateApkClient.updateApk(updateTask, TEST_URL, activity);
+    latch.await(1000, TimeUnit.MILLISECONDS);
+    updateApkClient.setInstallationResult(RESULT_FAILED);
+
+    assertFalse(updateTask.isSuccessful());
+    assertTrue(updateTask.getException() instanceof FirebaseAppDistributionException);
+    FirebaseAppDistributionException e =
+        (FirebaseAppDistributionException) updateTask.getException();
+    assertEquals("Installation failed with result code: " + RESULT_FAILED, e.getMessage());
+    assertEquals(FirebaseAppDistributionException.Status.INSTALLATION_FAILURE, e.getErrorCode());
+  }
+
+  @Test
+  public void downloadTask_whenCalledMultipleTimes_getsCancelled() {
```

Comment:
Added this question to the class under test too, but should this return the same download task instead of cancelling?